### DISCUSSION TITLE
Working with files

### DISCUSSION THREAD

**Discussion by rostidev** [helix-editor/helix#11204] [Ideas]
I've just discovered the Helix editor a few days ago and from a brief view it looks cool, useful and very promising to me. One of the cool things of Helix is its file picker. But at the same time the file picker has also disappointed me. I mean the idea of it is cool but at least my user experience was frustrated.

First of all the file picker doesn't have functionality of changing directory. It collects all files from either the current directory or the current Git worktree recursively and allows to choose a file from that list of files only. In my first run of Helix I ran it from its own directory (just double clicked on its executable) and was very wondering why I can't go to the parent directory from the file picker and edit some file that is not of Helix itself. Later I found that Helix supports changing the current directory by the `:change-current-directory` command, outside the file picker. But this is not convenient at all. Working with one big list of recursively found files is not convenient 300%! Look how it looks if the current directory is `C:\`

![image](https://github.com/user-attachments/assets/83094ba6-58cf-403a-8565-43aa38551d7b)

It also takes a long time to make that flat 611k+ list of files and the paths are truncated.

But it's cool anyway because if looks similarly to Norton Commander and to all other orthodox file managers (OFM) like [Far Manager](https://www.farmanager.com/screenshots.php?l=en), [Altap Salamander](https://www.altap.cz/), Midnight Commander, etc. They also divide the screen into left and right windows and also have the preview functionality of the selected file. For example:

![image](https://github.com/user-attachments/assets/cde68190-4f74-49e0-9ebd-2aef4b3faf32)

![image](https://github.com/user-attachments/assets/4d8902d0-176d-4ed5-859a-2ca4381b8603)

I propose to change the file picker to have a sub-set of that functionality of OFMs and to have a similar user experience. The minimum functionality that I expect to have is:
* Convenient way of walking through the file system hierarchy without full recursive scan and without the need to leave the file picker.
* Showing names of files and directories in the current directory without their full paths and not including files/directories of any sub-directory in the current list.
* Show full path of the current directory in some single place (in OFM this is the top side of the current window).
* Management of files and directories: view, edit, copy, move/rename, make directory, delete (F3, F4, F5, F6, F7, F8 in OFM).

In case you don't want to add functionality of changing disk (because its Windows only) you can use the same approach that is used by MSYS2 and Cygwin where Windows disk letters are just directories in the root. For example `/d/develop/jansi/src` instead of `D:\develop\jansi\src`.

**Comment by rostidev** [comment]
Showing a full recursively made list of all files under the current directory could also be an option but not the default option. I mean, scanning 611k+ files by default is a very bad idea and this should be done so only when a user really needs it and knows the list will not be huge.

**Reply by kirawi** [reply]
Opening files from `C:\` is not the intended use-case of the file picker. You are supposed to open Helix in the project directory and open the file picker there.

**Reply by rostidev** [reply]
> Opening files from `C:\` is not the intended use-case of the file picker. You are supposed to open Helix in the project directory and open the file picker there.

Of course. I used `C:\` just as an example of a big project and of opening Helix in such a place accidentally. Also projects could have many binary or generated files that the user doesn't want to edit and see in the list. Projects can also use some other VCS instead of Git and there could be an equivalent of that VCS's `.git`-like directory that Hellix doesn't know to filter out.

**Reply by kirawi** [reply]
The picker supports filtering syntax: https://github.com/junegunn/fzf?tab=readme-ov-file#search-syntax

Helix supports `.ignore` and `.helix/ignore` in the absence of `.gitignore` which should cover other VCSs.. Helix also ignores hidden files by default.

**Comment by kirawi** [comment]
The file picker works like `telescope.nvim`. It sounds like you are looking for a file tree, and this is not planned for core. There is already a `vim-dirvish`-like file tree implemented for the in-progress plugin system: https://github.com/helix-editor/helix/pull/8675 & https://github.com/mattwparas/helix-config/blob/master/cogs/file-tree.scm

**Reply by stibinator** [reply]
Telescope's file picker lets you navigate through the filesystem, where you can move to the parent directory with `-`. If the file picker worked like Telescope that would be great.

**Reply by kirawi** [reply]
That should be possible now on master, if I'm not mistaken. https://github.com/helix-editor/helix/pull/11285

**Reply by stibinator** [reply]
It doesn't seem to. If I hit `-` it just filters for files with '-' in their names. I tried a buncha different key combos and couldn't get it to go up a directory, am I missing something?

**Reply by dahyun-jeon-90** [reply]
> It doesn't seem to. If I hit `-` it just filters for files with '-' in their names. I tried a buncha different key combos and couldn't get it to go up a directory, am I missing something?

is there any progress on this issue?
I am experiencing same issue
